openshift/dynamic/apply.py

```diff
@@ -25,18 +25,22 @@ def apply(resource, definition):
         del actual_dict['metadata']['annotations'][LAST_APPLIED_CONFIG_ANNOTATION]
         patch = merge(last_applied, definition, actual_dict)
         if patch:
-            return resource.patch(body=dict_merge(patch, desired_annotation),
-                                  name=definition['metadata']['name'],
-                                  namespace=definition['metadata'].get('namespace'),
-                                  content_type='application/merge-patch+json')
+            return dict_merge(patch, desired_annotation)
         else:
             return actual
     else:
-        return resource.patch(
-            body=definition,
-            name=definition['metadata']['name'],
-            namespace=definition['metadata'].get('namespace'),
-            content_type='application/merge-patch+json')
+        return definition
+
+
+def apply(resource, definition):
+    definition = apply_object(resource, definition)
+    if isinstance(definition, dict):
```

Comment:
What is this check for? I'm not seeing how definition would stop being a dict

Reply:
It's for the `return actual` case of `apply_object`

Reply:
ah, `actual` is a `ResourceInstance` and not a `dict`, I see. 

I'm not sure I like relying on the return type for determining whether we issue a patch or not, how do you feel about returning a `(dict, bool)` tuple representing `(definition, changed)` from the `apply_object` function instead? 

Not a blocker for me, just a thought, so if you'd rather not I can go ahead and merge.

Reply:
It's probably more readable your way, which is the more important thing (the fact that you had to ask about this suggest there's room for improvement)

Reply:
ah, I'd forgotten about the fourth case (return `resource.create`) which further complicates matters - not sure the tuple helps there